pkgs/development/tools/java/java-language-server/default.nix

```diff
@@ -0,0 +1,100 @@
+{ lib, stdenv, fetchFromGitHub
+, jdk, maven, runtimeShell
+}:
+
+let
+  platform =
+    if stdenv.isLinux then "linux"
+    else if stdenv.isDarwin then "mac"
+    else if stdenv.isWindows then "windows"
+    else throw "unsupported platform";
+in
+stdenv.mkDerivation rec {
+  pname = "java-language-server";
+  version = "0.2.38";
+
+  src = fetchFromGitHub {
+    owner = "georgewfraser";
+    repo = pname;
+    # commit hash is used as owner sometimes forgets to set tags. See https://github.com/georgewfraser/java-language-server/issues/104
+    rev = "1dfdc54d1f1e57646a0ec9c0b3f4a4f094bd9f17";
+    sha256 = "sha256-zkbl/SLg09XK2ZhJNzWEtvFCQBRQ62273M/2+4HV1Lk=";
+  };
+
+  fetchedMavenDeps = stdenv.mkDerivation {
+    name = "java-language-server-${version}-maven-deps";
+    inherit src;
+    buildInputs = [ maven ];
+
+    buildPhase = ''
+      runHook preBuild
+
+      mvn package -Dmaven.repo.local=$out/.m2 -DskipTests
+
+      runHook postBuild
+    '';
+
+    installPhase = ''
+      runHook preInstall
+
+      find $out -type f \
+        -name \*.lastUpdated -or \
+        -name resolver-status.properties -or \
+        -name _remote.repositories \
+        -delete
+
+      runHook postInstall
+    '';
+
+    dontFixup = true;
+    dontConfigure = true;
+    outputHashAlgo = "sha256";
+    outputHashMode = "recursive";
+    outputHash = "sha256-2MVlF3QIWiDvUlnMH4RLi2Od57aoh8zK/OmHqztOnZ4=";
+  };
+
+
+  nativeBuildInputs = [ maven jdk ];
+
+  dontConfigure = true;
+  buildPhase = ''
+    runHook preBuild
+
+    jlink \
+      ${lib.strings.optionalString (!stdenv.isDarwin) "--module-path './jdks/${platform}/jdk-13/jmods'"} \
+      --add-modules java.base,java.compiler,java.logging,java.sql,java.xml,jdk.compiler,jdk.jdi,jdk.unsupported,jdk.zipfs \
+      --output dist/${platform} \
+      --no-header-files \
+      --no-man-pages \
+      --compress 2
+
+    mvn package --offline -Dmaven.repo.local=${fetchedMavenDeps}/.m2 -DskipTests
+
+    runHook postBuild
+  '';
+
+  installPhase = ''
+    runHook preInstall
+
+    mkdir -p $out/share/java/java-language-server
+    cp -r dist/classpath dist/*${platform}* $out/share/java/java-language-server
+
+    mkdir -p $out/bin
+    # a link is not used as lang_server_${platform}.sh makes use of "dirname $0" to access other files
+    cat << _EOF > $out/bin/java-language-server
+    #!${runtimeShell}
+    $out/share/java/java-language-server/lang_server_${platform}.sh "\$@"
+    _EOF
+    chmod +x $out/bin/java-language-server
```

Comment:
Yeah, I mentioned it in the comment in the code, but I guess I should expand. Suppose it was a symlink and consider this line in `lang_server_linux.sh` https://github.com/georgewfraser/java-language-server/blob/master/dist/launch_linux.sh#L11. According to my testing and this answer https://unix.stackexchange.com/a/234619, $0 would refer to the symlink and not the target file. Therefore, `dirname $0` would evalate to `/nix/store/...java-language-server/bin` instead of `...../java-language/server/share/java/java-language-server` and as you can see this would cause issues in the rest of the code.

Initially, I tried to symlink it but when I noticed this, I only saw two solutions, have a script like how it is done in the AUR or patch the files. I wasn't sure how to properly patch the files in a simple manner, so I resorted to a script. 

Please let me know if there is better way to solve this issue (perhaps using symlinks)

Reply:
Doh, I should have read more closely :).

*perhaps* [wrapProgram}(https://nixos.org/manual/nixpkgs/unstable/#fun-wrapProgram)/[makeWrapper](https://nixos.org/manual/nixpkgs/unstable/#fun-makeWrapper) might be slightly nicer, but that's a matter of taste.

Reply:
Thanks, I didnt know such a tool existed